### PR TITLE
Switch to using include directive for Nginx constants

### PR DESCRIPTION
**Pull request by ohemorange**
Addresses #4524.

### CODE REVIEW

**Comment by ohemorange** [comment]
This undoes parts of #2262 and #3383 (but leaves the parts about fixing the `ssl_session_cache`, for example).

We previously thought that it would be easier to upgrade ciphersuites if we kept them inline, but it turns out that's not the case.

Yes, this runs the risk of accidentally duplicating specific directives in a single server block. But, that's arguably safer now than it was when we made the original change. First, we only ever do this to non-ssl server blocks, and we more robustly detect which are and aren't server blocks now. So the only case where it's a problem is when there are already ssl directives in a non-ssl server block, at which point something else has probably also already gone even more terribly wrong.

Some people who used `/etc/letsencrypt/options-ssl-nginx.conf` in the past might have manually modified it, because it lives in `etc`. We should make sure that old versions are kept around when we add the updating code (which isn't in here just yet, and should probably just go in a separate PR). As a bonus for adding that, people who have the super-old version (from before we changed to `ssl_session_cache shared:le_nginx_SSL:1m;`) will get the update to fix that for free!

**Comment by ohemorange** [comment]
There's a slight benefit to doing it this way.

Previously, we tried to insert the directive, and if we found it at all, we told them and didn't bother trying to insert it and rolled everything back.

Now, we just stick it in (effectively) and try to start nginx. Because sometimes it's just a warning, not a failure.

So it's optimistically try to stick it in and start nginx anyway vs. go parse the file, test its contents for duplicates in the server block, and only if it all checks out stick it in.

Which gives us a chance of it succeeding vs. a better error message.

I think the error message is worth it, I'll change it here.

**Comment by ohemorange** [comment]
@bmw took some refactoring, but error message is back.

File: certbot-nginx/certbot_nginx/parser.py [hunk 1]
```diff
@@ -344,6 +312,16 @@ def add_server_directives(self, vhost, directives, replace):
         except errors.MisconfigurationError as err:
             raise errors.MisconfigurationError("Problem in %s: %s" % (filename, str(err)))
 
+def _parse_flat_file(ssl_options):
```
Comment (bmw) [inline certbot-nginx/certbot_nginx/parser.py]: nit: If you want to make this function be specific to our TLS options file, I'd update the function name to reflect that. I see a benefit in making this more general as we continue to use more include directives though so perhaps you should update variable names and error messages to make it more useful for parsing included files in the future.

File: certbot-nginx/certbot_nginx/parser.py [hunk 2]
```diff
@@ -536,8 +515,10 @@ def _add_directive(block, directive, replace):
 
     # Find the index of a config line where the name of the directive matches
     # the name of the directive we want to add. If no line exists, use None.
-    location = next((index for index, line in enumerate(block)
-                     if line and line[0] == directive[0]), None)
+    find_location = lambda direc: next((index for index, line in enumerate(block) \
```
Comment (bmw) [inline certbot-nginx/certbot_nginx/parser.py]: nit: perhaps you should make a small helper function for this as this is a bit complicated to read.

File: certbot-nginx/certbot_nginx/parser.py [hunk 3]
```diff
@@ -549,15 +530,36 @@ def _add_directive(block, directive, replace):
...
         # and there is already a copy of that directive with a different value
         # in the config file.
+
+
+        # handle flat include files
+
         directive_name = directive[0]
-        if location is None or (isinstance(directive_name, str) and
-                                directive_name in REPEATABLE_DIRECTIVES):
+        can_append = lambda loc, dir_name: loc is None or (isinstance(dir_name, str) \
```
Comment (bmw) [inline certbot-nginx/certbot_nginx/parser.py]: nit: consider splitting into small helper function rather than lambda.

File: certbot-nginx/certbot_nginx/parser.py [hunk 4]
```diff
@@ -549,15 +530,36 @@ def _add_directive(block, directive, replace):
...
+                                and dir_name in REPEATABLE_DIRECTIVES)
+
+        cannot_append_error = lambda direc, loc: errors.MisconfigurationError( \
```
Comment (bmw) [inline certbot-nginx/certbot_nginx/parser.py]: nit: This is a somewhat strange use of a lambda in my opinion. Maybe instead you should make an `ERR_FMT` string that you can use or a simple function called something like `_fail_add_directive()`.

**Comment by ohemorange** [comment]
Oops meant to preserve behavior, I'll add that condition back in.

File: certbot-nginx/certbot_nginx/parser.py (hunk 4 above)
Comment (ohemorange) [inline certbot-nginx/certbot_nginx/parser.py]: I may have gone a little overboard on the lambdas.

File: certbot-nginx/certbot_nginx/parser.py (hunk 1 above)
Comment (ohemorange) [inline certbot-nginx/certbot_nginx/parser.py]: given that was already have a complicated general mechanism for parsing files, I'm going to return this to its original name and leave it as a special-use function.

If #583 ever gets resolved, that would be a good time to fix our general machinery for parsing files and included files and figuring out which directives apply where.

**Comment by ohemorange** [comment]
Actually, there will be a behavior change here.

Before, if there was an exact duplicate, we wouldn't error out, but nor would we append it to have two copies of the same directive in the same block.

Now, it's probably most correct to include the entire file containing that directive and (as we currently do) just fall back to Nginx's own error checking to figure out whether that was ok or not on restart.

**Comment by bmw** [comment]
I want to do some more testing here but after playing with this for a bit, it appears whether or not Nginx errors out if there a duplicate directive depends on the directive. This is fine:
```
ssl_session_cache shared:le_nginx_SSL:1m;
ssl_session_cache shared:le_nginx_SSL:1m;
```
but this is not:
```
ssl_prefer_server_ciphers on;
ssl_prefer_server_ciphers on;
```
This is an unfortunate side effect of using `include` because previously we just wouldn't include the directive if an identical one was already present. This means that switching to this approach will cause Certbot to fail for configurations where it previously worked.

Unfortunately, I have to sign off for the evening, but I'll play with and think about this more tomorrow.

**Comment by ohemorange** [comment]
Yeah, that's right. We could comment out the original directive with a note saying "duplicated in <include file location>."On May 1, 2017 6:37 PM, Brad Warren <notifications@github.com> wrote:I want to do some more testing here but after playing with this for a bit, it appears whether or not Nginx errors out if there a duplicate directive depends on the directive. This is fine:
ssl_session_cache shared:le_nginx_SSL:1m;
ssl_session_cache shared:le_nginx_SSL:1m;

but this is not:
ssl_prefer_server_ciphers on;
ssl_prefer_server_ciphers on;

This is an unfortunate side effect of using include because previously we just wouldn't include the directive if an identical one was already present. This means that switching to this approach will cause Certbot to fail for configurations where it previously worked.
Unfortunately, I have to sign off for the evening, but I'll play with and think about this more tomorrow.

**Comment by bmw** [comment]
What would you like to see with this PR? Try to get functionality like that added in time for 0.14.0? Skip this for 0.14.0? Ship this knowing it will no longer work on some existing configurations?

**Comment by ohemorange** [comment]
That's like one extra line, I can do it in about 5 minutes. If you'd rather this wait, I'm always fine with that.

**Comment by bmw** [comment]
If it's that simple, go for it! It won't work if the directive is duplicated outside of the server block, but we have that same problem currently so it certainly doesn't have to be fixed in this PR.

**Comment by ohemorange** [comment]
Right yeah, if we want this whole shebang to be correct that would be more and different work.
